### PR TITLE
remove autotools build system

### PR DESCRIPTION
**Pull request by sezero**
- TODO: update INSTALL.txt to replace the autotools configure
        instructions with cmake.
- TODO: update make build system to provide an equivalent to
        autotools' `make dist` ?
- TODO: update / revise github actions, replace autotools-only
        ones with cmake (e.g.: vmactions.yml for FreeBSD.)

Reference issue: https://github.com/libsdl-org/SDL/issues/6571

### CODE REVIEW

**Comment by sezero** [comment]
@madebr: Feel free to revise the TODO items

**Comment by madebr** [comment]
I'ld like the removal to wait until after creation of binary/source packages has been added to CMake.

**Comment by slouken** [comment]
I'm not planning to add source and binary packaging to CMake.
For source packages I'll generate VERSION.txt in a clean directory and then archive it up.
For binary packages I have a pre-existing flow that moves installed products into place and adds documentation and platform specific notes. And Linux distribution maintainers have a similar flow using their tools.

**Comment by sezero** [comment]
> For source packages I'll generate VERSION.txt in a clean directory and then archive it up.

You need to do something for SDL_revision.h too.

You guys feel free to push to my `autofoo` branch for any revisions you have.

**Comment by madebr** [comment]
There is still a reference to the configure build system in `docs/README-raspberrypi.md` about cross building SDL from desktop to rpi.
I'm sure it works, but don't what are the correct commands so I left that file alone.

File: CMakeLists.txt [hunk 1]
```diff
@@ -2971,8 +2971,6 @@ configure_file("${SDL3_SOURCE_DIR}/sdl3.pc.in"
   "${SDL3_BINARY_DIR}/sdl3.pc" @ONLY)
 configure_file("${SDL3_SOURCE_DIR}/sdl3-config.in"
   "${SDL3_BINARY_DIR}/sdl3-config" @ONLY)
-configure_file("${SDL3_SOURCE_DIR}/SDL3.spec.in"
-  "${SDL3_BINARY_DIR}/SDL3.spec" @ONLY)
```
Comment (sezero) [inline CMakeLists.txt:2975]: OK, distros will rely on their on cmake'ified rpm spec files.

**Comment by sezero** [comment]
TODO's remaining:
- Edit `docs/README-raspberrypi.md` as mentioned by @madebr above
- Change CI:: vmactions.yml for FreeBSD to use cmake instead of autofoo
- General tidy-up of CI runners after autofoo removal
- Automate `SDL_revision.h` generation for source tarball distribution?
- Same for `VERSION.txt` ?

P.S.: I'll most certainly go to hell after this for sins of favoring cmake.

File: build-scripts/cmake-toolchain-mingw64-i686.cmake [hunk 2]
```diff
@@ -0,0 +1,13 @@
+set(CMAKE_SYSTEM_NAME Windows)
+set(CMAKE_SYSTEM_PROCESSOR x86)
+
+find_program(CMAKE_C_COMPILER NAMES i686-w64-mingw32-gcc)
+find_program(CMAKE_CXX_COMPILER NAMES i686-w64-mingw32-g++)
```
Comment (sezero) [inline build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: You also want a `CMAKE_RC_COMPILER` (same for 64 bit version.)

**Comment by slouken** [comment]
I have updated the release process to use CMake instead of autotools.

@sezero, I hear the autotools gods are benevolent and forgiving.

**Comment by sezero** [comment]
> @sezero, I hear the autotools gods are benevolent and forgiving.

I sure hope so. And I hope that the plain-makefile gods are benevolent too..

File: build-scripts/cmake-toolchain-mingw64-i686.cmake (hunk 2 above)
Comment (slouken) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: Interesting, I didn't end up needing that for some reason. @madebr?

Comment (madebr) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: Me neither.
But I guess it's a fail safe such that people know `CMAKE_RC_COMPILER` is needed for "something".
It's not an error to add it.

Comment (sezero) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: Is cmake finding it by itself based on a predicted prefix from the C compiler, or something?

Comment (sezero) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: > `CMAKE_RC_COMPILER` is needed for "something"

we need it for building version.rc, do we not?

Comment (madebr) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: Yes, it's for `version.rc`. (I was saying it in perspective of SDL users)

CMake sets a [default `CMAKE_RC_COMPILER` here](https://github.com/Kitware/CMake/blob/93ded7304aea9848352ce843d98fbf64023be11f/Modules/Platform/Windows-GNU.cmake#L159-L161).

The [cross compiler prefix is guessed here](https://github.com/Kitware/CMake/blob/93ded7304aea9848352ce843d98fbf64023be11f/Modules/CMakeDetermineCCompiler.cmake#L154-L193).

**Comment by madebr** [comment]
FreeBSD with CMake fails with:
```
2022-11-24T19:26:10.3313460Z In file included from /Users/runner/work/SDL/SDL/src/video/x11/SDL_x11events.c:33:
2022-11-24T19:26:10.3314760Z /Users/runner/work/SDL/SDL/src/video/x11/SDL_x11xinput2.h:30:36: error: typedef redefinition with different types ('struct XGenericEventCookie' vs 'struct XGenericEventCookie' (aka 'XGenericEventCookie'))
2022-11-24T19:26:10.3315830Z typedef struct XGenericEventCookie XGenericEventCookie;
2022-11-24T19:26:10.3316690Z                                    ^
2022-11-24T19:26:10.3317520Z /usr/local/include/X11/Xlib.h:967:3: note: previous definition is here
2022-11-24T19:26:10.3318340Z } XGenericEventCookie;
2022-11-24T19:26:10.3319050Z   ^
2022-11-24T19:26:10.3319780Z 1 error generated.
```

File: build-scripts/cmake-toolchain-mingw64-i686.cmake (hunk 2 above)
Comment (sezero) [reply build-scripts/cmake-toolchain-mingw64-i686.cmake:5]: OK. (Can't test it myself though -- my cmake is 3.9.6, mingw is made to require 3.13+)

**Comment by sezero** [comment]
> FreeBSD with CMake fails with:
> 
> ```
> 2022-11-24T19:26:10.3313460Z In file included from /Users/runner/work/SDL/SDL/src/video/x11/SDL_x11events.c:33:
> 2022-11-24T19:26:10.3314760Z /Users/runner/work/SDL/SDL/src/video/x11/SDL_x11xinput2.h:30:36: error: typedef redefinition with different types ('struct XGenericEventCookie' vs 'struct XGenericEventCookie' (aka 'XGenericEventCookie'))
> 2022-11-24T19:26:10.3315830Z typedef struct XGenericEventCookie XGenericEventCookie;
> 2022-11-24T19:26:10.3316690Z                                    ^
> 2022-11-24T19:26:10.3317520Z /usr/local/include/X11/Xlib.h:967:3: note: previous definition is here
> 2022-11-24T19:26:10.3318340Z } XGenericEventCookie;
> 2022-11-24T19:26:10.3319050Z   ^
> 2022-11-24T19:26:10.3319780Z 1 error generated.
> ```
> 
> Looking into it.

Cmake configuration reports:
```
  -- Performing Test HAVE_XGENERICEVENT
  -- Performing Test HAVE_XGENERICEVENT - Failed
```

... but autotools in main branch reports `   checking for XGenericEvent... yes`

Something isn't right.

**Comment by madebr** [comment]
I think I found it. The BSD's store the X include headers in special locations.
So we have to explicitly add these locations to the check paths.

**Comment by sezero** [comment]
```
  [314/583] /usr/bin/cc -DSDL3_EXPORTS -DUSING_GENERATED_CONFIG_H -I/Users/runner/work/SDL/SDL/build/include -I/Users/runner/work/SDL/SDL/build/include-config- -I/usr/local/include -I/usr/local/include/libdrm -I/usr/local/include/libepoll-shim -I/Users/runner/work/SDL/SDL/build/wayland-generated-protocols -I/usr/local/include/dbus-1.0 -I/usr/local/lib/dbus-1.0/include -I/usr/local/include/ibus-1.0 -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -idirafter "/Users/runner/work/SDL/SDL/src/video/khronos"   -Wall -fno-strict-aliasing -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fcolor-diagnostics -mmmx -m3dnow -msse -msse2 -msse3 -I/usr/local/include -I/usr/local/include/pipewire-0.3 -I/usr/local/include/spa-0.2 -D_REENTRANT -I/usr/local/include/libdrm -DHAVE_USBHID_H -DUSBHID_NEW -D_REENTRANT -D_THREAD_SAFE  -DSDL_BUILD_MAJOR_VERSION=3 -DSDL_BUILD_MINOR_VERSION=0 -DSDL_BUILD_MICRO_VERSION=0 -fPIC -MD -MT CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o -MF CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o.d -o CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o -c /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c
  FAILED: CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o 
  /usr/bin/cc -DSDL3_EXPORTS -DUSING_GENERATED_CONFIG_H -I/Users/runner/work/SDL/SDL/build/include -I/Users/runner/work/SDL/SDL/build/include-config- -I/usr/local/include -I/usr/local/include/libdrm -I/usr/local/include/libepoll-shim -I/Users/runner/work/SDL/SDL/build/wayland-generated-protocols -I/usr/local/include/dbus-1.0 -I/usr/local/lib/dbus-1.0/include -I/usr/local/include/ibus-1.0 -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -idirafter "/Users/runner/work/SDL/SDL/src/video/khronos"   -Wall -fno-strict-aliasing -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fcolor-diagnostics -mmmx -m3dnow -msse -msse2 -msse3 -I/usr/local/include -I/usr/local/include/pipewire-0.3 -I/usr/local/include/spa-0.2 -D_REENTRANT -I/usr/local/include/libdrm -DHAVE_USBHID_H -DUSBHID_NEW -D_REENTRANT -D_THREAD_SAFE  -DSDL_BUILD_MAJOR_VERSION=3 -DSDL_BUILD_MINOR_VERSION=0 -DSDL_BUILD_MICRO_VERSION=0 -fPIC -MD -MT CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o -MF CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o.d -o CMakeFiles/SDL3.dir/src/core/linux/SDL_evdev.c.o -c /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:619:23: error: use of undeclared identifier 'SDL_UDEV_DEVICE_TOUCHPAD'
          (udev_class & SDL_UDEV_DEVICE_TOUCHPAD) ? SDL_TOUCH_DEVICE_INDIRECT_ABSOLUTE : SDL_TOUCH_DEVICE_DIRECT,
                        ^
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:772:26: warning: implicit declaration of function 'NBITS' is invalid in C99 [-Wimplicit-function-declaration]
      unsigned long relbit[NBITS(REL_MAX)] = { 0 };
                           ^
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:772:26: error: variable-sized object may not be initialized
      unsigned long relbit[NBITS(REL_MAX)] = { 0 };
                           ^~~~~~~~~~~~~~
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:800:32: warning: implicit declaration of function 'test_bit' is invalid in C99 [-Wimplicit-function-declaration]
          item->relative_mouse = test_bit(REL_X, relbit) && test_bit(REL_Y, relbit);
                                 ^
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:806:23: error: use of undeclared identifier 'SDL_UDEV_DEVICE_TOUCHSCREEN'
      if (udev_class & (SDL_UDEV_DEVICE_TOUCHSCREEN | SDL_UDEV_DEVICE_TOUCHPAD)) {
                        ^
  /Users/runner/work/SDL/SDL/src/core/linux/SDL_evdev.c:806:53: error: use of undeclared identifier 'SDL_UDEV_DEVICE_TOUCHPAD'
      if (udev_class & (SDL_UDEV_DEVICE_TOUCHSCREEN | SDL_UDEV_DEVICE_TOUCHPAD)) {
                                                      ^
  2 warnings and 4 errors generated.
```

**Comment by madebr** [comment]
CMake:
```
```

**Comment by madebr** [comment]
We should perhaps add a warning banner to the SDL2 configure script informing people that it is EOL.

**Comment by sezero** [comment]
> We should perhaps add a warning banner to the SDL2 configure script informing people that it is EOL.

I hope not.

**Comment by slouken** [comment]
> We should perhaps add a warning banner to the SDL2 configure script informing people that it is EOL.

No, we'll leave it as-is.

**Comment by slouken** [comment]
@madebr, I tried using cpack to build the source archive, and I failed pretty hard, so I just do this instead:
```
    (cd $PRODUCT && build-scripts/updaterev.sh --dist) || exit 1
    rm -rf $PRODUCT-$VERSION
    cp -a $PRODUCT $PRODUCT-$VERSION
    rm -rf $PRODUCT-$VERSION/.git*
```

**Comment by slouken** [comment]
> Does anybody know where that file usually hangs out?

On Linux it's in /usr/include/linux/input.h
On FreeBSD it's in /usr/local/include/linux/input.h

In general the BSD ports system installs to /usr/local, so we should be looking there for native builds.

**Comment by sezero** [comment]
> In general the BSD ports system installs to /usr/local, so we should be looking there for native builds.

Isn't their compiler setup to look for headers under /usr/local/include ?

**Comment by slouken** [comment]
> > In general the BSD ports system installs to /usr/local, so we should be looking there for native builds.
> 
> Isn't their compiler setup to look for headers under /usr/local/include ?

No, I don't believe so.

**Comment by madebr** [comment]
I fixed building FreeBSD on github ci (and also pushed it onto main).
Because of the conflicts, I rebased this pr.

**Comment by slouken** [comment]
@madebr, are there any other discrepancies between the autotools generated config.h and the cmake generated one?

If everything else looks clean, I'm good with this being merged.

**Comment by madebr** [comment]
After comparing the headers, the following is missing from CMake:

- loongarch support: missing checks + lsx library + following defines:
    ```
    #undef HAVE_LSXINTRIN_H
    #undef HAVE_LASXINTRIN_H
    ```
- `SDL_INPUT_WSCONS` for OpenBSD wscons input driver
- `SDL_UDEV_DYNAMIC` for dynamic udev support

**Comment by madebr** [comment]
These discrepancies are addressed by #6615 and #6616.

**Comment by Conan-Kudo** [comment]
We should probably still test CMake with makefiles and ninja files. The builds run differently and SDL2 CMake is sufficiently custom that I'd prefer to see that tested.

**Review by slouken — APPROVED** [review]

**Review by Conan-Kudo — APPROVED** [review]

File: docs/README-riscos.md [hunk 3]
```diff
@@ -12,19 +12,13 @@ Requirements:
...
+Currently, SDL for RISC OS only supports compiling with GCCSDK under Linux.
 
 The following commands can be used to build SDL for RISC OS using CMake:
 
     cmake -Bbuild-riscos -DCMAKE_TOOLCHAIN_FILE=$GCCSDK_INSTALL_ENV/toolchain-riscos.cmake -DRISCOS=ON -DCMAKE_INSTALL_PREFIX=$GCCSDK_INSTALL_ENV -DCMAKE_BUILD_TYPE=Release
     cmake --build build-riscos
-    cmake --build build-riscos --target install
+    cmake --install build-riscos
 
 When using GCCSDK 4.7.4 release 6 or earlier versions, the builtin atomic functions are broken, meaning it's currently necessary to compile with `--disable-gcc-atomics` using autotools or `-DSDL_GCC_ATOMICS=OFF` using CMake. Newer versions of GCCSDK don't have this problem.
```
Comment (ccawley2011) [inline docs/README-riscos.md:23]:
```suggestion
When using GCCSDK 4.7.4 release 6 or earlier versions, the builtin atomic functions are broken, meaning it's currently necessary to compile with `-DSDL_GCC_ATOMICS=OFF` using CMake. Newer versions of GCCSDK don't have this problem.
```

File: docs/README-macos.md [hunk 4]
```diff
@@ -93,7 +82,7 @@ NSApplicationDelegate implementation:
 
 # Using the Simple DirectMedia Layer with a traditional Makefile
 
-An existing autoconf/automake build system for your SDL app has good chances
+An existing CMake build system for your SDL app has good chances
```
Comment (ccawley2011) [inline docs/README-macos.md:85]:
```suggestion
An existing build system for your SDL app has good chances

```

File: docs/README-macos.md [hunk 5]
```diff
@@ -38,9 +27,9 @@ script.
 ```bash
 mkdir build
 cd build
-CC=$PWD/../build-scripts/clang-fat.sh ../configure
-make
-sudo make install
+cmake .. "-DCMAKE_OSX_ARCHITECTURES=arm64;x86_64"
```
Comment (ccawley2011) [inline docs/README-macos.md:30]: You'll want to adapt the preceding description as well.

File: docs/README-porting.md [hunk 6]
```diff
@@ -11,14 +11,14 @@ on based on C preprocessor symbols.
 
 There are two basic ways of building SDL at the moment:
 
-1. The "UNIX" way:  ./configure; make; make install
+1. CMake:  cmake -S . -B build && cmake --build build && cmake --install install
 
-   If you have a GNUish system, then you might try this.  Edit configure.ac,
+   If you have a GNUish system, then you might try this.  Edit CMakeLists.txt,
```
Comment (ccawley2011) [inline docs/README-porting.md:16]:
```suggestion
   If you have a system that supports CMake, then you might try this.  Edit CMakeLists.txt,

```

**Comment by sezero** [comment]
Rebased, hopefully without loss.

What else is left for this?

**Comment by madebr** [comment]
Not essentially for this thread, but this still references removed files

- Update  `docs/README-raspberrypi.md` for cross building to rpi
- Update  `docs/README-macos.md` for creating a fat library.
  * `build-scripts/clang-fat.sh` is removed.
  * The file also uses `sdl-config` instead of `sdl3-config`

**Comment by sezero** [comment]
> Not essentially for this thread, but this still references removed files
> 
>     * Update  `docs/README-raspberrypi.md` for cross building to rpi
> 
>     * Update  `docs/README-macos.md` for creating a fat library.
>       
>       * `build-scripts/clang-fat.sh` is removed.
>       * The file also uses `sdl-config` instead of `sdl3-config`

@slouken ?

**Comment by slouken** [comment]
Yay!

**Comment by madebr** [comment]
Doing this on a friday. Courageous man you are :)

**Comment by Conan-Kudo** [comment]
Woohoo!

**Pull request merged**